tests/baselines/reference/generatorReturnTypeInference.types

```diff
@@ -0,0 +1,308 @@
+=== tests/cases/conformance/generators/generatorReturnTypeInference.ts ===
+declare const iterableIterator: IterableIterator<number>;
+>iterableIterator : IterableIterator<number>
+
+declare const generator: Generator<number, string, boolean>;
+>generator : Generator<number, string, boolean>
+
+declare const never: never;
+>never : never
+
+function* g000() { // Generator<never, void, unknown>
+>g000 : () => Generator<never, void, unknown>
+}
+
+// 'yield' iteration type inference
+function* g001() { // Generator<undefined, void, unknown>
+>g001 : () => Generator<undefined, void, unknown>
+
+    yield;
+>yield : any
+}
+
+function* g002() { // Generator<number, void, unknown>
+>g002 : () => Generator<number, void, unknown>
+
+    yield 1;
+>yield 1 : any

```

Comment:
This is because we fall back to `any` in `checkYieldExpression` if we don't have a return type annotation. I can make this `unknown`, but I could also possibly infer it from other yield expressions:

```ts
function *g() {
  let x: string = yield;
  let y = yield; // should `y` be `unknown` or `string`?
}
```

I need to verify the logic isn't circular, but it could be feasible to infer `string` for `y` by first inferring the return type of `g`, then inferring the result of `yield` from `TNext` of the inferred return type. At first glance it doesn't seem to be circular as we only infer `TNext` contextually from type annotations (so `x` is a candidate for contextual typing of `TNext`, but not `y`).

Reply:
In general that seems to work however I have an odd case of an escaping type parameter for this test case:

*generatorReturnTypeInference.ts*
```ts
declare function f2<T>(x: T): T;
function* g204() { // Generator<number, void, any>
    const x = f2(yield 1);
}
```

*generatorReturnTypeInference.types*
```
function* g204() { // Generator<number, void, any>
>g204 : () => Generator<number, void, T>

	const x = f2(yield 1);
>x : T
>f2(yield 1) : T
>f2 : <T>(x: T) => T
>yield 1 : T
>1 : 1
```

You end up with an undeclared `T` that comes from `f2`.

Reply:
Relying on other statements in the body sounds hard to get right. `any` is fine with me for now, and `undefined` even better. I just got confused with the variety of possible return types.

Reply:
Yeah, I've looked into it and there are a number of issues with trying to set up the machinery for that. `yield` returning `any` when there is no return type annotation is the best we can do currently.